### PR TITLE
[mlir][python] Add normalforms to capture preconditions of transforms

### PR DESCRIPTION
**Pull request by martin-luecke**
This adds the concept of a _Normalform_ to the transform dialect python extras.

Normalforms are defined by a sequence of transforms required to ensure this normalform. The term and design is inspired by a similar concept in the term rewriting community.

Each transform handle adheres to a normalform. Initially simply `Normalform`, the weakest normalform, that all other normalforms inherit from.
A normalform might imply other normalforms which is currently modeled using inheritance. i.e. stronger normalforms inherit from weaker normalforms.

Note that the normalform of a handle is a static property currently without precise info about the actual payload.

Normalizing a handle to a normalform might trigger propagation of this form to parent and child handles, depending on the specific normalform.
An example for a (conceived) normalform is the `PerfectForNestForm` that aims to achieve perfect loop nests in the IR, if possible, by using canonicalization and loop invariant code motion transforms.
e.g.
```python
class PerfectForNestForm(Normalform):
  propagate_up = False
  propagate_down = True

  def _impl(cls, handle: OpHandle) -> None:
    with handle.apply_patterns(): 
      structured.ApplyTilingCanonicalizationPatternsOp()
      loop.ApplyForLoopCanonicalizationPatternsOp() 
      transform.ApplyCanonicalizationPatternsOp()
    handle.apply_licm()
    handle.apply_cse()
```

This normalform is propagated only to child handles, as all handles to operations that are nested at a deeper level will have also been impacted by these transforms and consist of perfectly nested loops, if possible. This normalform is not propagated to parent handles as these are not impacted by this specific normalization.

With the current design Normalforms are never instantiated. Only ever a type of normalform is used, e.g.
```python
class Handle(ir.Value):
  @property
  def normalform(self) -> Type["Normalform"]:
```
This could possibly also be modeled as an `Enum`, but that makes modeling a hierarchy of normalforms more complicated.

### CODE REVIEW

File: mlir/python/mlir/dialects/transform/extras/__init__.py [hunk 1]
```diff
@@ -42,6 +44,46 @@ def __init__(
         super().__init__(v)
         self.parent = parent
         self.children = children if children is not None else []
+        self._normalform = Normalform
+
+    @property
+    def normalform(self) -> Type["Normalform"]:
```
Comment (makslevental) [inline mlir/python/mlir/dialects/transform/extras/__init__.py:50]: Just curious - I see this is techinically `typing.Type[ForwardRef('Normalform')]` but is this different from just writing `"Normalform"`?

File: mlir/python/mlir/dialects/transform/extras/__init__.py [hunk 2]
```diff
@@ -192,6 +234,45 @@ def constant_param(value: Union[ir.Attribute, int]) -> ParamHandle:
     return op.param
 
 
+class Normalform:
```
Comment (makslevental) [inline mlir/python/mlir/dialects/transform/extras/__init__.py:237]: ultra nit: prefer `NormalForm`

File: mlir/python/mlir/dialects/transform/extras/__init__.py [hunk 3]
```diff
@@ -42,6 +44,46 @@ def __init__(
...
+            self.parent.propagate_up_normalform(normalform)
+
+    def propagate_down_normalform(self, normalform: Type["Normalform"]):
+        for child in self.children:
+            # We set the child normalform directly to avoid infinite recursion
+            # in case this normalform needs to be propagated up and down.
+            child._normalform = normalform
+            child.propagate_down_normalform(normalform)
+
+    def normalize(self: "HandleT", normalform: Type["Normalform"]) -> "HandleT":
```
Comment (makslevental) [inline mlir/python/mlir/dialects/transform/extras/__init__.py:84]: `HandleT` is defined above so you can just use it here?

File: mlir/python/mlir/dialects/transform/extras/__init__.py [hunk 4]
```diff
@@ -192,6 +234,45 @@ def constant_param(value: Union[ir.Attribute, int]) -> ParamHandle:
...
+    propagate_down: bool = True
+
+    def __init__(self):
+        raise TypeError(
+            "Normalform cannot be instantiated directly. Use Type[Normalform]"
+            "instead."
+        )
+
+    @classmethod
+    def _impl(cls, handle: "HandleT") -> "HandleT":
```
Comment (makslevental) [inline mlir/python/mlir/dialects/transform/extras/__init__.py:257]: `HandleT`

**Review by makslevental — APPROVED** [review]
LGTM modulo nits

**Comment by joker-eph** [comment]
I'm confused, why is this a python concept?

**Comment by martin-luecke** [comment]
@joker-eph 
> I'm confused, why is this a python concept?

This does not have to be a Python concept. In fact, we thought about possibly tracking the
normalform of a handle through the MLIR type of that handle. 

For now, we opted to have this in a shallow frontend abstraction rather than in the dialect itself.
We want to get some mileage with this concept and see how it eases the generation of transform scripts.
That includes additional extensions to this that are harder to realize in transform core (e.g. requiring new passes or extensive changes to the transfrom core infrastructure).
For instance, I am working on a follow-up PR that introduces automatic normalization if a handle is not statically known to be in a specific normalform. The goal is to remove some of the guesswork of which canonicalizations / clean-up passes need to be applied before a specific transform.

Additionally, this design might change in the future as we learn more about it. For instance, I am not sure if modeling a hierarchy of normalforms using inheritance is strong enough to express the possible implications between a variety of normalforms in the same and across dialects.

Note that all of this is only concerned with the generation of a transform script. It does not do any analysis / transformations on existing transform IR.

In summary, we want to get some more experience with transform pre/post-conditions and introduce modeling of this in the dialect itself when we know better what aspects are key.

**Comment by joker-eph** [comment]
> In summary, we want to get some more experience with transform pre/post-conditions and introduce modeling of this in the dialect itself when we know better what aspects are key.

I'm just not sure this belongs to MLIR right now. I believe the python bindings are just that... bindings. That would be a significant departure to start adding "features" that don't exist in MLIR.

**Comment by martin-luecke** [comment]
The question is how much lifting the bindings are allowed to do to make IR generation easier. 
In my opinion, these are convenience extras to help generate transform IR more easily. 
No analysis or transformation of existing IR is happening here.

The goal is for instance to be able to write
```python
with Context(), InsertionPoint(body), Autonormalize(): 
  matmul = module.match_op(linalg.matmulOp)
  matmul.tile(tile_sizes=[4,8])
```
instead of:
```python
with Context(), InsertionPoint(body): 
  matmul = module.match_op(linalg.matmulOp)
  module.apply_tiling_canonicalization()
  module.apply_licm()
  module.canonicalize()
  matmul.tile(tile_sizes=[4,8])
```
to generate similar IR.

I think the Python bindings have slowly transitioned into more than bindings. Rather than simply exposing the C++ API to Python, they provide sensible pythonic frontend abstractions. 
For instance, we can annotate Python functions to generate `func.func` operations with an automatically inserted `func.return` for the SSA values returned from the _Python_ function, rather than calling into C++ APIs explicitly for this. 
AFAIK the C++ Dialect APIs are not concerned with easing IR emittance beyond OpBuilders. Abstractions that automatically inject ops are in Python only and improve usability there a lot.
Similar to other Python abstractions, I don't think this is designed as a feature that makes Python a non-optional component for transform dialect. 

The feedback about this at the LLVM devmtg was encouraging and to me, they did not feel out of place in this environment of "easing IR generation" so I started upstreaming.

Would you be happier with this feature if we had it in C++ as well or even found a way to encode this in the IR?

Of course, if there are objections to such extras I will take this down.

**Comment by martin-luecke** [comment]
@joker-eph ping

**Comment by joker-eph** [comment]
> The question is how much lifting the bindings are allowed to do to make IR generation easier.
> In my opinion, these are convenience extras to help generate transform IR more easily.

The "concept of normalform" is just a convenience? 
From reading this PR description this seems more than that to me, and it seems like a concept that could be purely native.

Ultimately to me the "bindings" are supposed to be just that: binding to the MLIR concepts. Of course there is some wrapping that can be done to make them more pythonic, safer, etc. But what you're doing here does not seem to fit this category of "let's make the MLIR API more pythonic" but really adding a new concept here.

**Comment by joker-eph** [comment]
@ftynse offered to setup meeting to chat about this if you'd like.
